zonemaster/zonemaster-backend
# Update database connection code

## Purpose

This PR updates the database connection code so that it's possible to get a database adapter instance that doesn't correspond to the contents of a configuration file. This is useful if you want to include database schema definition in the database adapter code.

This is mostly about refactoring, but it also includes fixing of a bug (introduced since last release) and some minor improvements.

## Context

This PR is a step towards both #805 and #806.

## Changes

* Fixes connecting to MySQL over IPv6.

* Fixes the syntax of the data source name for PostgreSQL. I'm not sure why the old syntax used to work or why it doesn't anymore (on CentOS 8), but this PR changes the syntax to the one specified in [DBD::Pg::connect](https://metacpan.org/pod/DBD::Pg#connect).

* The database adapter constructor interfaces are changed from `new({ config => $config })`to `from_config( $config )`. This makes room for adding other constructors that don't take their parameters straight out of a Config instance.

* There are a number of methods in the database adapters that used to look into the configuration for parameters. Now they are updated to take those parameters as arguments instead.

* Database adapters used to defer the database connection for as long as possible. However in order to avoid failures in inconvenient places we were explicitly working around that feature by forcing the opening of a database connection as early as possible. This PR removes the connection deferral feature.

* The DRY principle is applied to getting the database adapter class and to opening database connections. The consistency across adapters is improved regarding log messages and database connection settings, and it also helps us stay consistent.

## How to test this PR

1. Verify that Travis was able to run the tests using all database engines.
2. For each database engine:
    1. Configure Backend to use this database engine
    2. Restart zm-rpcapi and zm-testagent
    3. Run a test using zmtest and verify you get a result back.
